File: week-2/day-7-OOP/algo-caesar-cipher/js/caesarCipher.js
exports.caesarCipher = function (string, shift_key) {
    let result = [];
    let key = shift_key % 26;

    for (const char of string) {
        if (!/[a-zA-Z]/.test(char)) {
            // If the character is not an alphabet character, add it as is.
            result.push(char);
        } else if (/[a-z]/.test(char)) {
            // If it's a lowercase alphabet character, apply the cipher.
            result.push(lowerCase(char, key));
        } else if (/[A-Z]/.test(char)) {
            // If it's an uppercase alphabet character, apply the cipher.
            result.push(upperCase(char, key));
        }
    }
    return result.join(''); // Join the result array into a string.
};

const lowerCase = (letter, key) => {
    let newKey = letter.charCodeAt(0) + key;
    return newKey <= 122 ? String.fromCharCode(newKey) : String.fromCharCode(96 + (newKey % 122));
};

const upperCase = (letter, key) => {
    let newKey = letter.charCodeAt(0) + key;
    return newKey <= 90 ? String.fromCharCode(newKey) : String.fromCharCode(64 + (newKey % 90));
};

// Example usage:
const text = "Boy! What a string!";
const shift = 3;
const encryptedText = exports.caesarCipher(text, shift);
console.log(encryptedText); // Output: "Khoor, Zruog!"

File: week-2/day-7-OOP/algo-caesar-cipher/python/caesar_cipher.py
def caesar_cipher(string, shift_amount):
    result = []
    for letter in string:
        if not letter.isalpha():
            result.append(letter)
        elif letter.islower():
            result.append(lowerCase(letter, shift_amount % 26))
        elif letter.isupper():
            result.append(upperCase(letter, shift_amount % 26))

    return "".join(result)


def lowerCase(letter, key):
    newLetterCode = ord(letter) + key
    return chr(newLetterCode) if newLetterCode <= 122 else chr(96 + newLetterCode % 122)

def upperCase(letter, key):
    newLetterCode = ord(letter) + key
    return chr(newLetterCode) if newLetterCode <= 90 else chr(64 + newLetterCode % 90)

if __name__ == "__main__":
    print(caesar_cipher("Boy! What a string!", -5)) # "Wjt! Rcvo v nomdib!"